File: app/src/main/java/com/sayler/gina/adapter/DaysAdapter.java
/**
 * Created by sayler on 2016-11-22.
 * <p>
 * Copyright 2016 MiQUiDO <http://www.miquido.com/>. All rights reserved.
 */
package com.sayler.gina.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.futuremind.recyclerviewfastscroll.SectionTitleProvider;
import com.sayler.gina.R;
import com.sayler.gina.ui.TextUtils;
import com.sayler.gina.util.Constatns;
import com.timehop.stickyheadersrecyclerview.StickyRecyclerHeadersAdapter;
import entity.Day;

import java.util.List;

/**
 * @author sayler
 */
public class DaysAdapter extends BaseRecyclerViewAdapter<Day> implements SectionTitleProvider, StickyRecyclerHeadersAdapter<DaysAdapter.DaysHeaderViewHolder> {


  public DaysAdapter(Context context, List<Day> items) {
    super(context, items);
  }

  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    View view = inflater.inflate(R.layout.it_days, parent, false);
    return new DaysViewHolder(view, DaysAdapter.this);
  }

  @Override
  public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
    if (holder instanceof DaysViewHolder) {
      DaysViewHolder viewHolder = (DaysViewHolder) holder;

      Day day = items.get(position);
      viewHolder.setItem(day, position);
      //content
      String contentFull = day.getContent();
      String contentShort = TextUtils.truncateTo(contentFull, 150, "...");

      //date
      String date = day.getDate().toString(Constatns.DATA_PATTERN_DAY_NUMBER_DAY_OF_WEEK);

      viewHolder.contentShort.setText(contentShort);
      viewHolder.content.setText(contentFull);
      viewHolder.time.setText(date);

      //expand
      viewHolder.expanded = false;
      viewHolder.time.setOnClickListener(view -> {
        if (viewHolder.expanded) {
          viewHolder.contentShort.setVisibility(View.VISIBLE);
          viewHolder.content.setVisibility(View.GONE);
        } else {
          viewHolder.contentShort.setVisibility(View.GONE);
          viewHolder.content.setVisibility(View.VISIBLE);
        }
        viewHolder.expanded = !viewHolder.expanded;
      });
    }
  }


  @Override
  public String getSectionTitle(int position) {
    return items.get(position).getDate().toString(Constatns.DATE_PATTERN_YEAR_MONTH_2_LINES);
  }

  @Override
  public long getHeaderId(int position) {
    return Math.abs(items.get(position).getDate().toString().substring(0, 7).hashCode());
  }

  @Override
  public DaysHeaderViewHolder onCreateHeaderViewHolder(ViewGroup parent) {
    View view = inflater.inflate(R.layout.it_header, parent, false);
    return new DaysHeaderViewHolder(view, DaysAdapter.this);
  }

  @Override
  public void onBindHeaderViewHolder(DaysHeaderViewHolder holder, int position) {
    holder.title.setText(items.get(position).getDate().toString(Constatns.DATE_PATTERN_YEAR_MONTH));
  }

  static final class DaysViewHolder extends RecyclerViewHolderWithOnItemClick<Day> {

    @Bind(R.id.content)
    public TextView content;
    @Bind(R.id.contentShort)
    public TextView contentShort;
    @Bind(R.id.day)
    public TextView time;
    public boolean expanded = false;

    DaysViewHolder(final View view, final BaseRecyclerViewAdapter<Day> baseRecyclerViewAdapter) {
      super(view, baseRecyclerViewAdapter);
      ButterKnife.bind(this, view);
    }

  }

  static final class DaysHeaderViewHolder extends RecyclerViewHolderWithOnItemClick<Day> {

    @Bind(R.id.content)
    public TextView title;

    DaysHeaderViewHolder(final View view, final BaseRecyclerViewAdapter<Day> baseRecyclerViewAdapter) {
      super(view, baseRecyclerViewAdapter);
      ButterKnife.bind(this, view);
    }
  }
}

File: app/src/main/java/com/sayler/gina/adapter/DaysAdapter.kt
/**
 * Created by sayler on 2016-11-22.
 *
 *
 */
package com.sayler.gina.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import butterknife.Bind
import butterknife.ButterKnife
import com.futuremind.recyclerviewfastscroll.SectionTitleProvider
import com.sayler.gina.R
import com.sayler.gina.domain.IDay
import com.sayler.gina.ui.truncateTo
import com.sayler.gina.util.Constants
import com.timehop.stickyheadersrecyclerview.StickyRecyclerHeadersAdapter

/**
 * @author sayler
 */
class DaysAdapter(context: Context, items: List<IDay>) : BaseRecyclerViewAdapter<IDay>(context, items), SectionTitleProvider, StickyRecyclerHeadersAdapter<DaysAdapter.DaysHeaderViewHolder> {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = inflater.inflate(R.layout.it_days, parent, false)
        return DaysViewHolder(view, this@DaysAdapter)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is DaysViewHolder) {
            with(holder) {
                val day = items[position]
                setItem(day, position)
                //attachments icon
                day.attachments.size.let { attachmentsCount ->
                    attachmentsCountView.visibility = if (attachmentsCount > 0) {
                        attachmentsCountView.text = attachmentsCount.toString()
                        View.VISIBLE
                    } else
                        View.GONE
                }

                //content
                val contentFull = day.content
                val contentShort = contentFull?.truncateTo(200, "...")
                contentShortView.text = contentShort
                contentView.text = contentFull

                //date
                day.date.toString(Constants.DATA_PATTERN_DAY_NUMBER_DAY_OF_WEEK).let { date ->
                    timeView.text = date
                }

                //expand on time click
                timeView.setOnClickListener { _ ->
                    if (expanded) showShortView() else showFullView()
                }
                showShortView()
            }
        }
    }

    override fun getSectionTitle(position: Int): String {
        return items[position].date.toString(Constants.DATE_PATTERN_YEAR_MONTH_2_LINES)
    }

    override fun getHeaderId(position: Int): Long {
        return Math.abs(items[position].date.toString().substring(0, 7).hashCode()).toLong()
    }

    override fun onCreateHeaderViewHolder(parent: ViewGroup): DaysHeaderViewHolder {
        val view = inflater.inflate(R.layout.it_header, parent, false)
        return DaysHeaderViewHolder(view, this@DaysAdapter)
    }

    override fun onBindHeaderViewHolder(holder: DaysHeaderViewHolder, position: Int) {
        holder.title.text = items[position].date.toString(Constants.DATE_PATTERN_YEAR_MONTH)
    }

    private fun DaysViewHolder.showFullView() {
        contentShortView.visibility = View.GONE
        contentView.visibility = View.VISIBLE
        expanded = true
    }

    private fun DaysViewHolder.showShortView() {
        contentShortView.visibility = View.VISIBLE
        contentView.visibility = View.GONE
        expanded = false
    }

    class DaysViewHolder(view: View, baseRecyclerViewAdapter: BaseRecyclerViewAdapter<IDay>) : BaseRecyclerViewAdapter.RecyclerViewHolderWithOnItemClick<IDay>(view, baseRecyclerViewAdapter) {
        @Bind(R.id.contentText)
        lateinit var contentView: TextView
        @Bind(R.id.contentShort)
        lateinit var contentShortView: TextView
        @Bind(R.id.day)
        lateinit var timeView: TextView
        @Bind(R.id.attachmentsCount)
        lateinit var attachmentsCountView: TextView
        var expanded = false

        init {
            ButterKnife.bind(this, view)
        }
    }

    class DaysHeaderViewHolder(view: View, baseRecyclerViewAdapter: BaseRecyclerViewAdapter<IDay>) : BaseRecyclerViewAdapter.RecyclerViewHolderWithOnItemClick<IDay>(view, baseRecyclerViewAdapter) {
        @Bind(R.id.contentText)
        lateinit var title: TextView

        init {
            ButterKnife.bind(this, view)
        }
    }
}
